tools/libs/light/libxl_pci.c

```diff
@@ -18,13 +18,356 @@
 
 #include "libxl_internal.h"
 
+#include <libxenvchan.h>
+#include <xenpcid.h>
+
 #define PCI_BDF                "%04x:%02x:%02x.%01x"
 #define PCI_BDF_SHORT          "%02x:%02x.%01x"
 #define PCI_BDF_VDEVFN         "%04x:%02x:%02x.%01x@%02x"
 #define PCI_OPTIONS            "msitranslate=%d,power_mgmt=%d"
 #define PCI_BDF_XSPATH         "%04x-%02x-%02x-%01x"
 #define PCI_PT_QDEV_ID         "pci-pt-%02x_%02x.%01x"
 
+/*
+ *******************************************************************************
+ * xenpcid support start
+ *******************************************************************************
+ */
+struct vchan_state {
+    /* Server domain ID. */
+    libxl_domid domid;
+    /* XenStore path of the server with the ring buffer and event channel. */
+    char *xs_path;
+
+    struct libxenvchan *ctrl;
+    int select_fd;
+    /* receive buffer */
+    char *rx_buf;
+    size_t rx_buf_size; /* current allocated size */
+    size_t rx_buf_used; /* actual data in the buffer */
+};
+
+/* Returns 1 if path exists, 0 if not, ERROR_* (<0) on error. */
+static int xs_path_exists(libxl__gc *gc, const char *xs_path)
+{
+    int rc;
+    const char *dir;
+
+    rc = libxl__xs_read_checked(gc, XBT_NULL, xs_path, &dir);
+    if ( rc )
+        return rc;
+    if ( dir )
+        return 1;
+    return 0;
+}
+
+static libxl_domid vchan_find_pcid_server(libxl_ctx *ctx)
+{
+    GC_INIT(ctx);
+    char **domains;
+    unsigned int i, n;
+    libxl_domid domid = DOMID_INVALID;
+
+    domains = libxl__xs_directory(gc, XBT_NULL, "/local/domain", &n);
+    if ( !n )
+        goto out;
+
+    for (i = 0; i < n; i++) {
+        int d;
+
+        if ( sscanf(domains[i], "%d", &d) != 1 )
+            continue;
+        if ( xs_path_exists(gc, GCSPRINTF(XENPCID_XS_PATH, d)) > 0 )
+        {
+            /* Found the domain where the pcid server lives. */
+            domid = d;
+            break;
+        }
+    }
+
+out:
+    GC_FREE;
+    return domid;
+}
+
+static int vchan_init_client(libxl__gc *gc, struct vchan_state *state)
+{
+    state->domid = vchan_find_pcid_server(CTX);
+    if ( state->domid == DOMID_INVALID )
+    {
+        LOGE(ERROR, "Can't find vchan server");
+        return ERROR_NOTFOUND;
+    }
+
+    state->xs_path = strdup(GCSPRINTF(XENPCID_XS_PATH, state->domid));
+    state->ctrl = libxenvchan_client_init(CTX->lg, state->domid,
+                                          state->xs_path);
+    if ( !state->ctrl )
+    {
+        LOGE(ERROR, "Couldn't intialize vchan client");
+        return ERROR_FAIL;
+    }
+    state->select_fd = libxenvchan_fd_for_select(state->ctrl);
+    LOG(INFO, "Intialized vchan client, server at %s", state->xs_path);
+    return 0;
+}
+
+static struct vchan_state *vchan_get_instance(libxl__gc *gc)
```

Comment:
Can you confirm that this code will never will be executed in multi-threaded environment?